crates/red_knot_python_semantic/src/types.rs

```diff
@@ -302,13 +302,12 @@ fn declarations_ty<'db>(
     let declared_ty = if let Some(second) = all_types.next() {
         let mut builder = UnionBuilder::new(db).add(first);
         for other in [second].into_iter().chain(all_types) {
-            // Make sure not to emit spurious errors relating to `Type::Todo`,
-            // since we only infer this type due to a limitation in our current model.
-            //
-            // `Unknown` is different here, since we might infer `Unknown`
+            // `Unknown` needs special handling, since we might infer `Unknown`
             // for one of these due to a variable being defined in one possible
             // control-flow branch but not another one.
-            if !first.is_equivalent_to(db, other) && !first.is_todo() && !other.is_todo() {
```

Comment:
Checking for Type::Todo is not necessary anymore now with the changed `is_equivalent_to`. I'm unsure if the addition of `… || (first == Type::Unknown && other == Type::Unknown)` is correct here. There are no tests that fail if I remove it. And I struggled to write one.

Reply:
Here is one form of the test that fails without this check:

```py
# error: [unresolved-import]
# error: [unresolved-import]
from unknown import A, B

def flag() -> bool: ...

if flag():
    x: A
else:
    x: B
x = 1
```

It requires suppressing some unresolved imports in order to get something typed as `Unknown` that can be used in a declaration. And it's not entirely clear to me in this example that we _shouldn't_ error on these as incompatible declarations.

I think if we have this condition, we should have it for `Any` as well, they should be treated the same. So after https://github.com/astral-sh/ruff/pull/14742 lands, we could write the same test using `Any` annotations, without having to suppress any errors.

All that said, I am really not settled on the right form of this conflicting-declarations check. It may be that we should not have this check at all, and should just silently union the live declarations. It may be that we should use "gradual equivalence" for this check, where `Any` or `Unknown` would be compatible with any other declaration. I feel like I need to see how this works on more real codebases, and that will be difficult, since today's typed codebases don't generally have multiple declarations of the same symbol, since mypy and pyright don't allow it.

So I would suggest that we do the simplest thing for now and just remove this extra condition (and the comment), and allow the check to be as strict as possible, then relax it as we are pushed to do so by actual user needs in the future.